builder/vagrant/step_initialize_vagrant.go

```diff
@@ -0,0 +1,136 @@
+package vagrant
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"text/template"
+
+	"github.com/hashicorp/packer/helper/multistep"
+	"github.com/hashicorp/packer/packer"
+)
+
+type StepInitializeVagrant struct {
+	BoxName      string
+	BoxVersion   string
+	Minimal      bool
+	Template     string
+	SourceBox    string
+	OutputDir    string
+	SyncedFolder string
+	GlobalID     string
+}
+
+var DEFAULT_TEMPLATE = `Vagrant.configure("2") do |config|
+  config.vm.box = "{{.BoxName}}"
+  {{ if ne .SyncedFolder "" -}}
+  		config.vm.synced_folder "{{.SyncedFolder}}", "/vagrant"
+  {{- else -}}
+  		config.vm.synced_folder ".", "/vagrant", disabled: true
+  {{- end}}
+end`
+
+type VagrantfileOptions struct {
+	SyncedFolder string
+	BoxName      string
+}
+
+func (s *StepInitializeVagrant) createInitializeCommand() (string, error) {
+	tplPath := filepath.Join(s.OutputDir, "packer-vagrantfile-template.erb")
+	templateFile, err := os.Create(tplPath)
+	templateFile.Chmod(0777)
+	if err != nil {
+		retErr := fmt.Errorf("Error creating vagrantfile %s", err.Error())
+		return "", retErr
+	}
+
+	var tpl *template.Template
+	if s.Template == "" {
+		// Generate vagrantfile template based on our default
+		tpl = template.Must(template.New("VagrantTpl").Parse(DEFAULT_TEMPLATE))
+	} else {
+		// Read in the template from provided file.
+		tpl, err = template.ParseFiles(s.Template)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	opts := &VagrantfileOptions{
+		SyncedFolder: s.SyncedFolder,
+		BoxName:      s.SourceBox,
+	}
+
+	err = tpl.Execute(templateFile, opts)
+	if err != nil {
+		return "", err
+	}
+
+	abspath, err := filepath.Abs(tplPath)
+	if err != nil {
+		return "", err
+	}
+
+	return abspath, nil
+}
+
+func (s *StepInitializeVagrant) prepInitArgs() ([]string, error) {
+	// Prepare arguments
+	initArgs := []string{}
+
+	if s.BoxName != "" {
+		initArgs = append(initArgs, s.BoxName)
+	}
+
+	initArgs = append(initArgs, s.SourceBox)
+
+	if s.BoxVersion != "" {
+		initArgs = append(initArgs, "--box-version", s.BoxVersion)
+	}
+
+	if s.Minimal {
+		initArgs = append(initArgs, "-m")
+	}
+
+	tplPath, err := s.createInitializeCommand()
+	if err != nil {
+		return initArgs, err
+	}
+
+	initArgs = append(initArgs, "--template", tplPath)
+
+	return initArgs, nil
+}
+
+func (s *StepInitializeVagrant) Run(_ context.Context, state multistep.StateBag) multistep.StepAction {
+	driver := state.Get("driver").(VagrantDriver)
+	ui := state.Get("ui").(packer.Ui)
+
+	// Skip the initialize step if we're trying to launch from a global ID.
+	if s.GlobalID != "" {
+		ui.Say("Using a global-id; skipping Vagrant init in this directory...")
+		return multistep.ActionContinue
+	}
+
+	ui.Say("Initializing Vagrant in build directory...")
+
+	initArgs, err := s.prepInitArgs()
+	if err != nil {
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+
+	os.Chdir(s.OutputDir)
```

Comment:
Won't we want to come back to the path we were before ? Could this be causing issues down the line, with a post-processor/else ?

Ideally, a `os.Chdir` shouldn't be required and the driver would work from where we are currently running, but I'm guessing you put this here for a reason 🙂. Just wondering why.

Reply:
Vagrant does a lot of scoping based on what directory you're in. I suppose I could push the directory changes to the driver itself, though.

Reply:
Ah, nice ! 🙂 Yes I think because they - processors and builders - run in different goroutines it is okay. This builder will change its directory, but not the others.

Reply:
oh, and I forgot, don't forget error management !  chdirs can fail 🙂

Reply:
Another option could be to use the `VAGRANT_CWD` environment variable.

> ####  `VAGRANT_CWD`
> `VAGRANT_CWD` can be set to change the working directory of Vagrant. By default, Vagrant uses the current directory you are in. The working directory is important because it is where Vagrant looks for the Vagrantfile. It also defines how relative paths in the Vagrantfile are expanded, since they're expanded relative to where the Vagrantfile is found.
> 
> This environmental variable is most commonly set when running Vagrant from a scripting environment in order to set the directory that Vagrant sees.

https://www.vagrantup.com/docs/other/environmental-variables.html#vagrant_cwd

Reply:
Yes, this is an easy way to not need to worry about tracking directories (the `VAGRANT_CWD` env var).

Reply:
That is super cool !!! Thanks for sharing.
Note that in go, you can add env vars only for a command by setting [`cmd.Env`](https://golang.org/pkg/os/exec/#Cmd)